src/gluonts/model/n_beats/_network.py

```diff
@@ -0,0 +1,531 @@
+        F, context_length, prediction_length, fwd_looking=is_forecast
+    )
+    cosines = F.stack(
+        *[F.cos(2 * np.pi * i * t) for i in range(num_coefficients)]
+    )
+    sines = F.stack(
+        *[F.sin(2 * np.pi * i * t) for i in range(num_coefficients)]
+    )
+    S = F.concat(cosines, sines, dim=0)
+    return F.dot(thetas, S)
+
+
+def trend_model(
+    F,
+    thetas: Tensor,
+    num_coefficients: int,
+    context_length: int,
+    prediction_length: int,
+    is_forecast: bool,
+) -> Tensor:
+    """
+    Creates a polynomial basis of degree num_coefficients-1.
+    """
+    t = linear_space(
+        F, context_length, prediction_length, fwd_looking=is_forecast
+    )
+    T = F.stack(*[t ** i for i in range(num_coefficients)])
+    return F.dot(thetas, T)
+
+
+class NBEATSBlock(mx.gluon.HybridBlock):
+    """
+    The NBEATS Block as described in the paper: https://arxiv.org/abs/1905.10437.
+    Its configurable to any of the specific block
+    types through the block_type parameter.
+
+    Parameters
+    ----------
+    width
+        Width of the fully connected layers with ReLu activation.
+    num_block_layers
+        Number of fully connected layers with ReLu activation.
+    expansion_coefficient_length
+        If the type is "G" (generic), then the length of the expansion coefficient.
+        If type is "T" (trend), then it corresponds to the degree of the polynomial.
+        If the type is "S" (seasonal) then its not used.
+    block_type
+        Either "G" (generic), "S" (seasonal) or "T" (trend).
+    prediction_length
+        Length of the prediction. Also known as 'horizon'.
+    context_length
+        Number of time units that condition the predictions
+        Also known as 'lookback period'.
+    has_backcast
+        Only the last block of the network doesnt.
+    kwargs
+        Arguments passed to 'HybridBlock'.
+    """
+
+    # Needs the validated decorator so that arguments types are checked and
+    # the block can be serialized.
+    @validated()
+    def __init__(
+        self,
+        width: int,
+        num_block_layers: int,
+        expansion_coefficient_length: int,
+        prediction_length: int,
+        context_length: int,
+        block_type: str,
+        has_backcast: bool,
+        **kwargs,
+    ) -> None:
+        super().__init__(**kwargs)
+
+        self.width = width
+        self.num_block_layers = num_block_layers
+        self.expansion_coefficient_length = expansion_coefficient_length
+        self.block_type = block_type
+        self.prediction_length = prediction_length
+        self.context_length = context_length
+        self.has_backcast = has_backcast
+
+        with self.name_scope():
+            self.fc_stack = mx.gluon.nn.HybridSequential()
+            for i in range(num_block_layers):
+                self.fc_stack.add(
+                    mx.gluon.nn.Dense(units=self.width, activation="relu")
+                )
+            if block_type == "G":
```

Comment:
As mentioned in the comment above, maybe instead of having this switching mechanism here you could have different block subtypes.